### PR TITLE
[4.0] Move template files to tmpl folder

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
With the move to namespaces, we didn't touch the location of the component template files, eg. _default.php_. They are still in a folder like _/components/com_content/views/article/tmpl_. This location was valid when we had the _view.html.php_ in the folder above. Now the class is in a different folder. So I think the template files should be moved to a location where it makes more sense and is inline with the rest of the core like modules or plugins.

This pr moves the template files to a **tmpl** folder within the component root folder below the view name. So _/components/com_content/views/article/tmpl/default.php_ becomes _/components/com_content/tmpl/article/default.php_.

The menu manager was adopted to support that new folder scheme as well as the view.

This pr was inspired by #15221 and is fully BC as all the other views should render as before. If this pr get accepted, we need to move all the template files of all the namespaced components.

### Testing Instructions
- Open an article in the back end.
- Create a new menu item with a "Single Article"
- Create a template override in the template manager for the article view and another one.

### Expected result
All should work as before. Also the template overrides should work.

### Actual result
Nothing is broken all should work.

### Documentation Changes Required
This new location needs to be documented.

### CODE REVIEW

**Comment by brianteeman** [comment]
wont the code in com_templates that creates overrides need changing as well?

**Comment by laoneo** [comment]
Nope, overrides are the same.

**Comment by mbabker** [comment]
It would need to know where to look to find the overrides though.

**Comment by dgt41** [comment]
@mbabker is there any plan to refactor the override creator part?

**Comment by laoneo** [comment]
I'v just tested it and overrides do work as before.

**Comment by dgt41** [comment]
@laoneo creating an override?

**Comment by laoneo** [comment]
Yes

**Comment by brianteeman** [comment]
> It would need to know where to look to find the overrides though.

how does it know where to look if the files have moved

**Comment by dgt41** [comment]
@brianteeman everything happens here:
https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_templates/models/template.php#L737-L786

So yes it should work as is

**Comment by laoneo** [comment]
Why do you guys don't believe me, it works :weary:

**Comment by brianteeman** [comment]
@dgt41 that code is for creating the files and folders - i am talking about the code that finds the views that can be overriden

**Comment by mbabker** [comment]
https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_templates/models/template.php#L569-L597

**Comment by brianteeman** [comment]
thanks @mbabker

**Comment by brianteeman** [comment]
and reading that code I cant see how it can find a view in the new location - but thats probably my lack of skill at reading code if @laoneo says it does. Anyway this is what I was referring to in my first comment

**Comment by laoneo** [comment]
This change here https://github.com/joomla/joomla-cms/pull/16218/files#diff-c0ed342062e86e3c3f81da894f9d3e8fR137 adds the tmpl path to the file lookup array.

**Comment by joomdonation** [comment]
@laoneo I think Brian is talking about the code which is used to create override in Template Manager. See the block of code which Michael mentioned https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_templates/models/template.php#L569-L597

Since the layout is now moved to new location, we need to adjust that block of code to allow template manager to find layout from new location

**Comment by laoneo** [comment]
I was thinking I do miss something. Will have a look on that as well.

**Comment by brianteeman** [comment]
Yes that's exactly what I meant

**Comment by laoneo** [comment]
Changed the override editor to support the new scheme as well. Updated the test instructions.

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on b9b0e6d1bd8988c87c7629f5d77cd9c97f4d68c4<br /><br />creating overrides is not building the path correctly if the site is installed in a subdirectory as it is including the subdirectory that the site is installed in<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16218">issues.joomla.org/tracker/joomla-cms/16218</a>.</sub>

**Comment by brianteeman** [comment]
this screenshot shows an override created before applying this pr at the bottom and after applying this pr at the top. you will see that my site is installed in a subdirectory called cms4


screenshotr10-46-56

**Comment by laoneo** [comment]
I just tested it and it again and I'm working in a subfolder as well, it worked as it should. For a test can you install it from here https://github.com/Digital-Peak/joomla-cms/archive/j4/move-views-tmpl.zip, just to be sure the full patch get applied.

**Comment by brianteeman** [comment]
Retested as requested with your zip - no change

screenshotr19-32-19

**Comment by laoneo** [comment]
Perhaps somebody else can test it as I don't know what to do more. Guess we will have another look on the code sprint as well. Thanks so far for testing.

**Comment by joomdonation** [comment]
I don't see the problem like Brian mentioned, but there are still issues with creating overrider:

1. For com_users (old layout structure), the override is created at html\\**view**\\com_users\\**views**\\login . It should be html\com_users\login only

2. For com_content (new layout structure), the override is created at html\\**view**\com_content\\**tmpl**\article, it should be html\com_content\article only

**Comment by joomdonation** [comment]
Forgot to mention I am using Windows. And I installed your branch for testing.  So something is still not correct yet

**Comment by brianteeman** [comment]
@joomdonation I am on windows too

File: administrator/components/com_templates/Model/Template.php [hunk 1]
```diff
@@ -575,30 +575,38 @@ public function getOverridesList()
...
+						}
+
+						// Check if the folder exists
+						if (!is_dir($path))
 						{
-							$result['components'][$component][] = $this->getOverridesFolder($view, $viewPath);
+							continue;
 						}
+
+						$result['components'][$component][] = $this->getOverridesFolder($view, $folder . '/');
```
Comment (joomdonation) [inline administrator/components/com_templates/Model/Template.php]: Change this line to
```php
$result['components'][$component][] = $this->getOverridesFolder($view, \JPath::clean($folder . '/'));
```
fixed the problem which I reported (on windows)

**Comment by joomdonation** [comment]
Also, maybe the code in foreach block of getOverridesList method could be changed to:

```php
foreach ($components as $component)
{
	$requireTmplFolder = true;

	if (file_exists($componentPath . '/' . $component . '/tmpl/'))
	{
		$viewPath = \JPath::clean($componentPath . '/' . $component . '/tmpl/');
		$requireTmplFolder = false;
	}
	elseif (file_exists($componentPath . '/' . $component . '/views/'))
	{
		$viewPath = \JPath::clean($componentPath . '/' . $component . '/views/');
	}
	elseif (file_exists($componentPath . '/' . $component . '/view/'))
	{
		$viewPath = \JPath::clean($componentPath . '/' . $component . '/view/');
	}
	else
	{
		$viewPath = '';
	}
	
	if ($viewPath)
	{
		$views = \JFolder::folders($viewPath);

		foreach ($views as $view)
		{
			// Only show the view has layout inside it
			if (!$requireTmplFolder || file_exists($viewPath . $view . '/tmpl'))
			{
				$result['components'][$component][] = $this->getOverridesFolder($view, $viewPath);
			}
		}
	}
}
```

I think it is less change, easier to understand (maybe just me)  and save \JFolder::folders call on unnecessary folders.

**Comment by brianteeman** [comment]
the override creation works for me now<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16218">issues.joomla.org/tracker/joomla-cms/16218</a>.</sub>

**Comment by laoneo** [comment]
Good to hear

**Comment by wilsonge** [comment]
Seems good to me. Let's get the rest of the core component converted :)

**Pull request merged**